crates/starknet-devnet-core/src/error.rs

```diff
@@ -189,3 +191,472 @@ pub enum MessagingError {
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct InnerContractExecutionError {
+    pub contract_address: starknet_api::core::ContractAddress,
+    pub class_hash: Felt,
+    pub selector: Felt,
+    #[serde(skip)]
+    pub return_data: Retdata,
+    pub error: Box<ContractExecutionError>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum ContractExecutionError {
+    /// Nested contract call stack trace frame.
+    Nested(InnerContractExecutionError),
+    /// Terminal error message.
+    Message(String),
+}
+
+impl From<String> for ContractExecutionError {
+    fn from(value: String) -> Self {
+        ContractExecutionError::Message(value)
+    }
+}
+
+use blockifier::execution::call_info::{CallInfo, Retdata};
+use serde::{Deserialize, Serialize};
+
+impl From<TransactionExecutionError> for ContractExecutionError {
+    fn from(value: TransactionExecutionError) -> Self {
+        let error_string = value.to_string();
+        let error_stack = gen_tx_execution_error_trace(&value);
+
+        (error_stack, error_string).into()
+    }
+}
+
+impl From<(blockifier::execution::stack_trace::ErrorStack, String)> for ContractExecutionError {
+    fn from(
+        (error_stack, error_string): (blockifier::execution::stack_trace::ErrorStack, String),
+    ) -> Self {
+        let mut recursive_error_option = Option::<ContractExecutionError>::None;
+        fn format_error(stringified_error: &str, error_cause: &str) -> String {
+            if stringified_error.is_empty() {
+                error_cause.to_string()
+            } else {
+                format!("{} {}", stringified_error, error_cause)
+            }
+        }
+
+        // [[[error inner] error outer] root]
+
+        for frame in error_stack.stack.iter().rev() {
+            let stack_err = match frame {
+                ErrorStackSegment::Cairo1RevertSummary(revert_summary) => {
+                    let mut recursive_error = ContractExecutionError::Message(format_error(
+                        &error_string,
+                        &serde_json::to_string(&revert_summary.last_retdata.0).unwrap_or_default(),
+                    ));
+
+                    for trace in revert_summary.stack.iter().rev() {
+                        recursive_error =
+                            ContractExecutionError::Nested(InnerContractExecutionError {
+                                contract_address: trace.contract_address,
+                                class_hash: trace.class_hash.unwrap_or_default().0,
+                                selector: trace.selector.0,
+                                return_data: revert_summary.last_retdata.clone(),
+                                error: Box::new(recursive_error),
+                            });
+                    }
+
+                    recursive_error
+                }
+                ErrorStackSegment::Vm(vm) => recursive_error_option.take().unwrap_or(
```

Comment:
I assume this is a way to skip the Vm frame because it gets assigned to itself again, it would be good to have a comment explaining this.